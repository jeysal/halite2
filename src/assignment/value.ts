import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';

import Goal from '../goal/goal';

const VALUE_IMPACT = 0.5;
const SUITABILITY_IMPACT = 0.5;

export interface ShipGoalValue<Turn> {
  ship: OwnShip<Turn>;
  goal: Goal<Turn>;
  val: number;
}

const calcShipGoalValue = <Turn>(
  ship: OwnShip<Turn>,
  goal: Goal<Turn>,
  gm: GameMap<Turn>,
): ShipGoalValue<Turn> => ({
  ship,
  goal,
  val:
    VALUE_IMPACT * goal.value(gm) +
    SUITABILITY_IMPACT * goal.suitability(ship, gm),
});

/**
 * generates a 2D matrix of ship-goal values with suitability taken into account
 */
const generateShipGoalValues = <Turn>(
  goals: Goal<Turn>[],
  gm: GameMap<Turn>,
): ShipGoalValue<Turn>[][] =>
  gm.myShips.map(ship => goals.map(goal => calcShipGoalValue(ship, goal, gm)));

export default generateShipGoalValues;
