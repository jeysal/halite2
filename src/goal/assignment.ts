import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';
import { maxBy } from 'ramda';

import Goal from '../goal/goal';

interface ShipGoal<Turn> {
  ship: OwnShip<Turn>;
  goal: Goal<Turn>;
}

const VALUE_IMPACT = 0.5;
const SUITABILITY_IMPACT = 0.5;

const shipGoalValue = <Turn>(
  ship: OwnShip<Turn>,
  goal: Goal<Turn>,
  gm: GameMap<Turn>,
): number =>
  VALUE_IMPACT * goal.value(gm) +
  SUITABILITY_IMPACT * goal.suitability(ship, gm);

const assignGoalsToShips = <Turn>(
  goals: Goal<Turn>[],
  gm: GameMap<Turn>,
): ShipGoal<Turn>[] => {
  const ownShips = gm.myShips;

  const shipGoalValues = ownShips.map(ship =>
    goals.map(goal => ({ ship, goal, val: shipGoalValue(ship, goal, gm) })),
  );

  const goalsWithShips = shipGoalValues.reduce((gwss, gVals) => {
    const gValsBelowLimit = gVals.filter(gVal => {
      const entry = gwss.find(gws => gws.goal === gVal.goal);
      return !entry || entry.ships.length < entry.goal.maxShips();
    });
    if (gValsBelowLimit.length) {
      const bestVal = gValsBelowLimit.reduce(maxBy(value => value.val));
      return gwss.map(
        goalWithShips =>
          bestVal && goalWithShips.goal === bestVal.goal
            ? {
                goal: goalWithShips.goal,
                ships: [...goalWithShips.ships, bestVal.ship],
              }
            : goalWithShips,
      );
    }
    return gwss;
  }, goals.map(goal => ({ goal, ships: [] as OwnShip<Turn>[] })));

  // TODO goal minima

  return goalsWithShips.reduce(
    (acc, gws) => [
      ...acc,
      ...gws.ships.map(ship => ({ ship, goal: gws.goal })),
    ],
    [] as ShipGoal<Turn>[],
  );
};

export default assignGoalsToShips;
