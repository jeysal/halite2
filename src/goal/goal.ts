import { Action } from 'halite/Game';
import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';

export default interface Goal<Turn = number> {
  value(gm: GameMap<Turn>): number;
  minShips(): number;
  maxShips(): number;
  suitability(ship: OwnShip<Turn>, gm: GameMap<Turn>): number;
  strive(ship: OwnShip<Turn>, gm: GameMap<Turn>): string & Action<Turn>;
};
