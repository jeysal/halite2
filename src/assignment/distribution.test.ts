import Goal from '../goal/goal';
import { goalStub } from '../test/goal';

import distributeShips from './distribution';
import { ShipGoalValue } from './value';

test('matches up the ships with their highest-value goals', () => {
  const goals: Goal[] = [{ ...goalStub }, { ...goalStub }, { ...goalStub }];

  expect(
    distributeShips([
      [
        { goal: goals[0], shipId: 0, val: 0.1 },
        { goal: goals[1], shipId: 0, val: 0.2 },
        { goal: goals[2], shipId: 0, val: 0.7 },
      ],
      [
        { goal: goals[0], shipId: 1, val: 0.6 },
        { goal: goals[1], shipId: 1, val: 0.5 },
        { goal: goals[2], shipId: 1, val: 0.4 },
      ],
      [
        { goal: goals[0], shipId: 2, val: 0.4 },
        { goal: goals[1], shipId: 2, val: 0.5 },
        { goal: goals[2], shipId: 2, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 3, val: 0.1 },
        { goal: goals[1], shipId: 3, val: 0.2 },
        { goal: goals[2], shipId: 3, val: 0.3 },
      ],
    ] as ShipGoalValue[][]).map(({ goal, shipIds }) => ({
      shipIds,
      goalIndex: goals.indexOf(goal),
    })),
  ).toMatchSnapshot();
});

test('does not match a ship that has value zero everywhere', () => {
  const goals: Goal[] = [{ ...goalStub }, { ...goalStub }];

  expect(
    distributeShips([
      [
        { goal: goals[0], shipId: 0, val: 0.5 },
        { goal: goals[1], shipId: 0, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 1, val: 0 },
        { goal: goals[1], shipId: 1, val: 0 },
      ],
    ] as ShipGoalValue[][]).map(({ goal, shipIds }) => ({
      shipIds,
      goalIndex: goals.indexOf(goal),
    })),
  ).toMatchSnapshot();
});

test('matches a ship to another goal to avoid exceeding maxShips', () => {
  const goals: Goal[] = [
    { ...goalStub },
    { ...goalStub },
    { ...goalStub, maxShips: () => 2 },
  ];

  expect(
    distributeShips([
      [
        { goal: goals[0], shipId: 0, val: 0.4 },
        { goal: goals[1], shipId: 0, val: 0.4 },
        { goal: goals[2], shipId: 0, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 1, val: 0.4 },
        { goal: goals[1], shipId: 1, val: 0.4 },
        { goal: goals[2], shipId: 1, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 2, val: 0.4 },
        { goal: goals[1], shipId: 2, val: 0.4 },
        { goal: goals[2], shipId: 2, val: 0.5 },
      ],
    ] as ShipGoalValue[][]).map(({ goal, shipIds }) => ({
      shipIds,
      goalIndex: goals.indexOf(goal),
    })),
  ).toMatchSnapshot();
});

test('does not match a ship if all goals are at capacity', () => {
  const goals: Goal[] = [
    { ...goalStub, maxShips: () => 1 },
    { ...goalStub, maxShips: () => 1 },
  ];

  expect(
    distributeShips([
      [
        { goal: goals[0], shipId: 0, val: 0.5 },
        { goal: goals[1], shipId: 0, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 1, val: 0.5 },
        { goal: goals[1], shipId: 1, val: 0.5 },
      ],
      [
        { goal: goals[0], shipId: 2, val: 0.5 },
        { goal: goals[1], shipId: 2, val: 0.5 },
      ],
    ] as ShipGoalValue[][]).map(({ goal, shipIds }) => ({
      shipIds,
      goalIndex: goals.indexOf(goal),
    })),
  ).toMatchSnapshot();
});

test('matches a ship to another goal to avoid undershooting minShips', () => {
  const goals: Goal[] = [
    { ...goalStub, minShips: () => 1 },
    { ...goalStub, minShips: () => 1 },
    { ...goalStub, minShips: () => 3 },
  ];

  expect(
    distributeShips([
      [
        { goal: goals[0], shipId: 0, val: 1 },
        { goal: goals[1], shipId: 0, val: 0 },
        { goal: goals[2], shipId: 0, val: 0 },
      ],
      [
        { goal: goals[0], shipId: 1, val: 0 },
        { goal: goals[1], shipId: 1, val: 0.3 },
        { goal: goals[2], shipId: 1, val: 0.7 },
      ],
      [
        { goal: goals[0], shipId: 2, val: 0 },
        { goal: goals[1], shipId: 2, val: 0.3 },
        { goal: goals[2], shipId: 2, val: 0.7 },
      ],
    ] as ShipGoalValue[][]).map(({ goal, shipIds }) => ({
      shipIds,
      goalIndex: goals.indexOf(goal),
    })),
  ).toMatchSnapshot();
});
