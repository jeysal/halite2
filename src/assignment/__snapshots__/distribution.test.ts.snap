// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not match a ship if all goals are at capacity 1`] = `
Array [
  Object {
    "goalIndex": 0,
    "shipIds": Array [
      0,
    ],
  },
  Object {
    "goalIndex": 1,
    "shipIds": Array [
      1,
    ],
  },
]
`;

exports[`does not match a ship the has value zero everywhere 1`] = `
Array [
  Object {
    "goalIndex": 0,
    "shipIds": Array [
      0,
    ],
  },
  Object {
    "goalIndex": 1,
    "shipIds": Array [],
  },
]
`;

exports[`matches a ship to another goal to avoid exceeding maxShips 1`] = `
Array [
  Object {
    "goalIndex": 0,
    "shipIds": Array [
      2,
    ],
  },
  Object {
    "goalIndex": 1,
    "shipIds": Array [],
  },
  Object {
    "goalIndex": 2,
    "shipIds": Array [
      0,
      1,
    ],
  },
]
`;

exports[`matches up the ships with their highest-value goals 1`] = `
Array [
  Object {
    "goalIndex": 0,
    "shipIds": Array [
      1,
    ],
  },
  Object {
    "goalIndex": 1,
    "shipIds": Array [
      2,
    ],
  },
  Object {
    "goalIndex": 2,
    "shipIds": Array [
      0,
      3,
    ],
  },
]
`;
